README.md
fakeetc
=======

[![Gem Version](http://img.shields.io/gem/v/fakeetc.svg?style=flat-square)][gem]
[![Build Status](http://img.shields.io/travis/sometimesfood/fakeetc.svg?style=flat-square)][travis]
[![Code Climate](http://img.shields.io/codeclimate/github/sometimesfood/fakeetc.svg?style=flat-square)][codeclimate]

A fake Ruby `Etc` module for your tests.

Intended as a drop-in replacement for [Etc][etc] in unit tests.

[etc]: http://ruby-doc.org/stdlib-2.2.0/libdoc/etc/rdoc/Etc.html

Usage
-----

```ruby
require 'fakeetc'

FakeEtc.add_groups({
  'empty' => { gid: 42, mem: [] },
  'anonymous' => { gid: 43, mem: ['johndoe'] }
})
FakeEtc.add_users({
  'janedoe' => { uid: 10,
                 gid: 20,
                 gecos: 'Jane Doe',
                 dir: '/home/janedoe',
                 shell: '/bin/zsh' },
  'jackdoe' => { uid: 50,
                 gid: 60,
                 gecos: 'Jack Doe',
                 dir: '/home/jackdoe',
                 shell: '/bin/bash' },
})

anonymous = nil
jack = nil

FakeEtc do
  anonymous = Etc.getgrnam('anonymous')
  jack = Etc.getpwuid(50)
end

anonymous
# => #<struct Struct::Group
#      name="anonymous",
#      passwd="x",
#      gid=43,
#      mem=["johndoe"]>

jack
# => #<struct Struct::Passwd
#      name="jackdoe",
#      passwd="x",
#      uid=50,
#      gid=60,
#      gecos="Jack Doe",
#      dir="/home/jackdoe",
#      shell="/bin/bash",
#      change=0,
#      uclass="",
#      expire=0>
```

Copyright
---------

Copyright (c) 2015 Sebastian Boehm. See LICENSE for details.

[gem]: https://rubygems.org/gems/fakeetc
[travis]: https://travis-ci.org/sometimesfood/fakeetc
[codeclimate]: https://codeclimate.com/github/sometimesfood/fakeetc


NEWS
* fakeetc NEWS
** 0.2.0 (2015-02-13)

 - Add support for user functions (FakeEtc.getgrgid, FakeEtc.getgrnam, ...)
 - Add documentation (yardoc)

** 0.1.0 (2015-02-02)

 - First public release

# Local Variables:
# mode: org
# End:


spec/fakeetc_spec.rb
require_relative 'spec_helper'

require 'fakeetc'

describe FakeEtc do
  before(:each) do
    @groups = {
      'cheeses' => { gid: 10, mem: %w(red_leicester tilsit) },
      'parrots' => { gid: 20, mem: %w(norwegian_blue macaw) }
    }
    @users = {
      'norwegian_blue' => { uid: 10,
                            gid: 20,
                            gecos: 'Remarkable Bird',
                            dir: '/home/parrot',
                            shell: '/bin/zsh' },
      'red_leicester' => { uid: 20,
                           gid: 10,
                           gecos: 'a little Red Leicester',
                           dir: '/home/red_leicester',
                           shell: '/bin/bash' }
    }
    FakeEtc.add_groups(@groups)
    FakeEtc.add_users(@users)
  end

  after(:each) do
    FakeEtc.endpwent
    FakeEtc.clear_users
    FakeEtc.endgrent
    FakeEtc.clear_groups
  end

  describe 'activate' do
    it 'should replace Etc with FakeEtc' do
      real_etc = Etc
      fake_etc = FakeEtc
      Etc.must_equal real_etc
      FakeEtc.activate
      Etc.must_equal fake_etc
      FakeEtc.deactivate
      Etc.must_equal real_etc
    end
  end

  describe 'with' do
    it 'should activate Etc in a block' do
      real_etc = Etc
      fake_etc = FakeEtc

      FakeEtc.with do
        Etc.must_equal fake_etc
      end
      Etc.must_equal real_etc
    end
  end

  describe 'without' do
    it 'should activate Etc in a block' do
      real_etc = Etc
      fake_etc = FakeEtc

      FakeEtc.with do
        Etc.must_equal fake_etc
        FakeEtc.without do
          Etc.must_equal real_etc
        end
        Etc.must_equal fake_etc
      end
      Etc.must_equal real_etc
    end
  end

  describe 'getgrnam' do
    it 'should find groups by name' do
      cheese_group = FakeEtc.getgrnam('cheeses')
      cheese_group.must_be_instance_of Struct::Group
      cheese_group.gid.must_equal @groups['cheeses'][:gid]
      cheese_group.mem.must_equal @groups['cheeses'][:mem]
    end

    it 'should raise exceptions for non-existent groups' do
      group_name = 'not-a-group'
      err = -> { FakeEtc.getgrnam(group_name) }.must_raise ArgumentError
      err.message.must_match "can't find group for #{group_name}"
    end
  end

  describe 'getgrgid' do
    it 'should find groups by gid' do
      parrot_group = FakeEtc.getgrgid(@groups['parrots'][:gid])
      parrot_group.must_be_instance_of Struct::Group
      parrot_group.gid.must_equal @groups['parrots'][:gid]
      parrot_group.mem.must_equal @groups['parrots'][:mem]
    end

    it 'should raise exceptions for non-existent groups' do
      gid = 247
      err = -> { FakeEtc.getgrgid(gid) }.must_raise ArgumentError
      err.message.must_match "can't find group for #{gid}"
    end
  end

  describe 'getgrent' do
    it 'should return all group entries in order' do
      cheese_group = FakeEtc.getgrent
      parrot_group = FakeEtc.getgrent
      nil_group = FakeEtc.getgrent

      cheese_group.name.must_equal 'cheeses'
      parrot_group.name.must_equal 'parrots'
      nil_group.must_be_nil
    end
  end

  describe 'endgrent' do
    it 'should reset group traversal' do
      cheese_group_1 = FakeEtc.getgrent
      FakeEtc.endgrent
      cheese_group_2 = FakeEtc.getgrent
      cheese_group_1.must_equal cheese_group_2
      cheese_group_1.name.must_equal 'cheeses'
    end
  end

  describe 'group' do
    it 'should execute a given block for each group entry' do
      groups = {}
      FakeEtc.group { |g| groups[g.name] = { gid: g.gid, mem: g.mem } }
      groups.must_equal @groups
    end

    it 'should behave like getgrent if there was no block given' do
      cheese_group = FakeEtc.group
      parrot_group = FakeEtc.group
      nil_group = FakeEtc.group

      cheese_group.name.must_equal 'cheeses'
      parrot_group.name.must_equal 'parrots'
      nil_group.must_be_nil
    end
  end

  describe 'sysconfdir' do
    it 'should call RealEtc.sysconfdir' do
      FakeEtc.sysconfdir.must_equal RealEtc.sysconfdir
    end
  end

  describe 'systmpdir' do
    it 'should call RealEtc.systmpdir' do
      FakeEtc.systmpdir.must_equal RealEtc.systmpdir
    end
  end

  describe 'getpwnam' do
    it 'should find users by name' do
      red_leicester = FakeEtc.getpwnam('red_leicester')
      red_leicester.must_be_instance_of Struct::Passwd
      red_leicester.uid.must_equal @users['red_leicester'][:uid]
      red_leicester.gid.must_equal @users['red_leicester'][:gid]
      red_leicester.gecos.must_equal @users['red_leicester'][:gecos]
      red_leicester.dir.must_equal @users['red_leicester'][:dir]
      red_leicester.shell.must_equal @users['red_leicester'][:shell]
    end

    it 'should raise exceptions for non-existent users' do
      user_name = 'not-a-user'
      err = -> { FakeEtc.getpwnam(user_name) }.must_raise ArgumentError
      err.message.must_match "can't find user for #{user_name}"
    end
  end

  describe 'getpwuid' do
    it 'should find users by uid' do
      norwegian_blue = FakeEtc.getpwuid(@users['norwegian_blue'][:uid])
      norwegian_blue.must_be_instance_of Struct::Passwd
      norwegian_blue.uid.must_equal @users['norwegian_blue'][:uid]
      norwegian_blue.gid.must_equal @users['norwegian_blue'][:gid]
      norwegian_blue.gecos.must_equal @users['norwegian_blue'][:gecos]
      norwegian_blue.dir.must_equal @users['norwegian_blue'][:dir]
      norwegian_blue.shell.must_equal @users['norwegian_blue'][:shell]
    end

    it 'should raise exceptions for non-existent groups' do
      uid = 247
      err = -> { FakeEtc.getpwuid(uid) }.must_raise ArgumentError
      err.message.must_match "can't find user for #{uid}"
    end
  end

  describe 'getpwent' do
    it 'should return all user entries in order' do
      norwegian_blue = FakeEtc.getpwent
      red_leicester = FakeEtc.getpwent
      nil_user = FakeEtc.getpwent

      norwegian_blue.name.must_equal 'norwegian_blue'
      red_leicester.name.must_equal 'red_leicester'
      nil_user.must_be_nil
    end
  end

  describe 'endpwent' do
    it 'should reset user traversal' do
      norwegian_blue_1 = FakeEtc.getpwent
      FakeEtc.endpwent
      norwegian_blue_2 = FakeEtc.getpwent
      norwegian_blue_1.must_equal norwegian_blue_2
      norwegian_blue_1.name.must_equal 'norwegian_blue'
    end
  end

  describe 'passwd' do
    it 'should execute a given block for each group entry' do
      users = {}
      FakeEtc.passwd do |u|
        users[u.name] = { uid: u.uid,
                          gid: u.gid,
                          gecos: u.gecos,
                          dir: u.dir,
                          shell: u.shell }
      end
      users.must_equal @users
    end

    it 'should behave like getpwent if there was no block given' do
      norwegian_blue = FakeEtc.passwd
      red_leicester = FakeEtc.passwd
      nil_user = FakeEtc.passwd

      norwegian_blue.name.must_equal 'norwegian_blue'
      red_leicester.name.must_equal 'red_leicester'
      nil_user.must_be_nil
    end
  end
end

describe 'FakeEtc' do
  it 'should activate FakeEtc and run a given block' do
    real_etc = Etc
    fake_etc = FakeEtc

    FakeEtc do
      Etc.must_equal fake_etc
    end
    Etc.must_equal real_etc
  end
end


lib/fakeetc/base.rb
# Most functions in this file are based on those in Chris Wanstrath's
# excellent fakefs library (lib/fakefs/base.rb).
#
# The following is a verbatim copy of the original license:
#
#   Copyright (c) 2009 Chris Wanstrath
#
#   Permission is hereby granted, free of charge, to any person obtaining
#   a copy of this software and associated documentation files (the
#   "Software"), to deal in the Software without restriction, including
#   without limitation the rights to use, copy, modify, merge, publish,
#   distribute, sublicense, and/or sell copies of the Software, and to
#   permit persons to whom the Software is furnished to do so, subject to
#   the following conditions:
#
#   The above copyright notice and this permission notice shall be
#   included in all copies or substantial portions of the Software.
#
#   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
#   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
#   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
#   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
#   LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
#   OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
#   WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

RealEtc = Etc

module FakeEtc # rubocop:disable Documentation
  @groups = {}
  @users = {}

  # Checks whether FakeEtc is currently activated.
  # @return [Bool] true if FakeEtc is currently activated, false if it
  #   is not
  def self.activated?
    @activated
  end

  # Activates FakeEtc.
  # @return [void]
  def self.activate
    @activated = true
    Object.class_eval do
      remove_const :Etc
      const_set :Etc, FakeEtc
    end
  end

  # Deactivates FakeEtc.
  # @return [void]
  def self.deactivate
    Object.class_eval do
      remove_const :Etc
      const_set :Etc, RealEtc
    end
    @activated = false
  end

  # Runs a code block with FakeEtc.
  # @return [Object] the block's return value
  def self.with
    if activated?
      yield
    else
      begin
        activate
        yield
      ensure
        deactivate
      end
    end
  end

  # Runs a code block without FakeEtc.
  # @return [Object] the block's return value
  def self.without
    if !activated?
      yield
    else
      begin
        deactivate
        yield
      ensure
        activate
      end
    end
  end
end

# Runs a code block with FakeEtc.
#
# @example
#   FakeEtc.add_groups('foo' => { gid: 42, mem: %w(bar baz) })
#   FakeEtc do
#     Etc.getgrnam('foo').gid
#   end
#   # => 42
#
# @return [Object] the block's return value
def FakeEtc(&block) # rubocop:disable Style/MethodName
  return ::FakeEtc unless block
  ::FakeEtc.with(&block)
end


lib/fakeetc/groups.rb
module FakeEtc # rubocop:disable Documentation
  class << self
    # Adds groups to the FakeEtc group list.
    #
    # @param group_hash [Hash{String=>Hash{Symbol=>Integer,String}}]
    #   the list of groups that should be added
    #
    # @example
    #   FakeEtc.add_groups({
    #     'empty' => { gid: 42, mem: [] },
    #     'anonymous' => { gid: 43, mem: ['johndoe'] }
    #   })
    #
    # @return [void]
    def add_groups(group_hash)
      passwd = 'x'
      group_hash.each do |group_name, group_info|
        group = Struct::Group.new(group_name,
                                  passwd,
                                  group_info[:gid],
                                  group_info[:mem])
        @groups[group_name] = group
      end
    end

    # Clears the group list.
    # @return [void]
    def clear_groups
      @groups = {}
    end

    # Finds a group by its group name.
    # @param group_name [String] the group's name
    # @return [Struct::Group] the group
    # @raise [ArgumentError] if no group with the given name can be
    #   found
    def getgrnam(group_name)
      group = @groups[group_name]
      fail ArgumentError, "can't find group for #{group_name}" if group.nil?
      group
    end

    # Finds a group by its gid.
    # @param gid [Integer] the group's gid
    # @return [Struct::Group] the group
    # @raise [ArgumentError] if no group with the given gid can be
    #   found
    def getgrgid(gid)
      group = @groups.values.find { |g| g.gid == gid }
      fail ArgumentError, "can't find group for #{gid}" if group.nil?
      group
    end

    # Returns an entry from the group list. Each successive call
    # returns the next entry or `nil` if the end of the list has been
    # reached.
    #
    # To reset scanning the group list, use {endgrent}.
    #
    # @return [Struct::Group] the next entry in the group list
    def getgrent
      @grents ||= @groups.values
      @grents.shift
    end

    # Ends the process of scanning through the group list.
    # @return [void]
    def endgrent
      @grents = nil
    end
    alias_method :setgrent, :endgrent

    # Executes a block for each group entry.
    # @yield [Struct::Group] the group entry
    # @return [void]
    def group
      return getgrent unless block_given?
      @groups.values.each { |g| yield g }
      endgrent
      nil
    end
  end
end


lib/fakeetc/users.rb
module FakeEtc # rubocop:disable Documentation
  class << self
    # Adds users to the FakeEtc user list.
    #
    # @param user_hash [Hash{String=>Hash{Symbol=>Integer,String}}]
    #   the list of users that should be added
    #
    # @example
    #   FakeEtc.add_users({
    #     'janedoe' => { uid: 10,
    #                    gid: 20,
    #                    gecos: 'Jane Doe',
    #                    dir: '/home/janedoe',
    #                    shell: '/bin/zsh' },
    #     'jackdoe' => { uid: 50,
    #                    gid: 60,
    #                    gecos: 'Jack Doe',
    #                    dir: '/home/jackdoe',
    #                    shell: '/bin/bash' },
    #   })
    #
    # @return [void]
    def add_users(user_hash)
      passwd = 'x'

      user_hash.each do |user_name, user_info|
        user = Struct::Passwd.new(user_name,
                                  passwd,
                                  user_info[:uid],
                                  user_info[:gid],
                                  user_info[:gecos],
                                  user_info[:dir],
                                  user_info[:shell])
        @users[user_name] = user
      end
    end

    # Clears the user list.
    # @return [void]
    def clear_users
      @users = {}
    end

    # Finds a user by their user name.
    # @param user_name [String] the user's name
    # @return [Struct::Passwd] the user
    # @raise [ArgumentError] if no user with the given name can be
    #   found
    def getpwnam(user_name)
      user = @users[user_name]
      fail ArgumentError, "can't find user for #{user_name}" if user.nil?
      user
    end

    # Finds a user by their user id.
    # @param uid [Integer] the user's id
    # @return [Struct::Passwd] the user
    # @raise [ArgumentError] if no user with the given id can be found
    def getpwuid(uid)
      user = @users.values.find { |u| u.uid == uid }
      fail ArgumentError, "can't find user for #{uid}" if user.nil?
      user
    end

    # Returns an entry from the user list. Each successive call
    # returns the next entry or `nil` if the end of the list has been
    # reached.
    #
    # To reset scanning the user list, use {endpwent}.
    #
    # @return [Struct::Passwd] the next entry in the user list
    def getpwent
      @pwents ||= @users.values
      @pwents.shift
    end

    # Ends the process of scanning through the user list.
    # @return [void]
    def endpwent
      @pwents = nil
    end
    alias_method :setpwent, :endpwent

    # Executes a block for each user entry.
    # @yield [Struct::Passwd] the user entry
    # @return [void]
    def passwd
      return getpwent unless block_given?
      @users.values.each { |u| yield u }
      endpwent
      nil
    end
  end
end
